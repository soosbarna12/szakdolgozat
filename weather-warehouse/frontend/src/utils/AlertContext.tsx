import React, { createContext, useState, useContext, ReactNode } from "react";
import { Snackbar, Alert } from "@mui/material";

type AlertContextType = {
    showAlert: (message: string, severity?: "error" | "info" | "success" | "warning") => void;
};

export const AlertContext = createContext<AlertContextType>({
    showAlert: () => { },
});

export function AlertProvider({ children }: { children: ReactNode }) {
    const [open, setOpen] = useState(false);
    const [message, setMessage] = useState("");
    const [severity, setSeverity] = useState<"error" | "info" | "success" | "warning">("info");

    function showAlert(newMessage: string, newSeverity: "error" | "info" | "success" | "warning" = "info") {
        setMessage(newMessage);
        setSeverity(newSeverity);
        setOpen(true);
    }

    return (
        <AlertContext.Provider value={{ showAlert }}>
            {children}
            <Snackbar open={open} autoHideDuration={4000} onClose={() => setOpen(false)}>
                <Alert severity={severity} onClose={() => setOpen(false)}>
                    {message}
                </Alert>
            </Snackbar>
        </AlertContext.Provider>
    );
}

export const useAlert = () => useContext(AlertContext);