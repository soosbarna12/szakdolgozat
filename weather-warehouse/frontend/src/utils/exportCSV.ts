import { download, generateCsv, mkConfig } from "export-to-csv";
import { HistoricalDataTable } from "../types/historicalDataTable.type";

export function exportCSV(data: HistoricalDataTable[]) {
  if (!data || data.length === 0) {
    return;
  }

  const csvConfig = mkConfig({ useKeysAsHeaders: true, filename: "historical_data" });

  const csv = generateCsv(csvConfig)(data as any[]);
  download(csvConfig)(csv);
}