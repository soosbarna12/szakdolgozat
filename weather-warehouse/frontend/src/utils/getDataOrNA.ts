export function getDataOrNA(data: any) {
  if (data) {
    return data;
  }
  return "N/A";
}