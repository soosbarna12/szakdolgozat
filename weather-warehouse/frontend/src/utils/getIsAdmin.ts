import { jwtDecode } from "jwt-decode";
import { JwtPayload } from "../types/jwtPayload";


export function getIsAdmin(): boolean {
    const token = localStorage.getItem('token');
    if (!token) return false;
    try {
      const decoded = jwtDecode<JwtPayload>(token);
      return decoded.role === 'admin';
    } catch {
      return false;
    }
  }