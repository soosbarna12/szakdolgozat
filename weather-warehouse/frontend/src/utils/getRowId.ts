import { HistoricalDataTable } from "../types/historicalDataTable.type";

export function getRowId(row: HistoricalDataTable): string {
  return row.date + row.cityName + row.countryCode;
}