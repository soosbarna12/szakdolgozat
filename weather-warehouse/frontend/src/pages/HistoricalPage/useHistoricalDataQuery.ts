import { useQuery } from "@tanstack/react-query";
import axios from "axios";
import { useAlert } from "../../utils/AlertContext";
import { UseHistoricalDataQueryProps } from "./HistoricalPage.type";
import { useEffect } from "react";

export function useHistoricalDataQuery(props: Readonly<UseHistoricalDataQueryProps>) {
  const { showAlert } = useAlert();
  const { location, date } = props;


  // Fetch weather data for the selected location
  const { data, error } = useQuery({
    queryKey: ["weather", location],
    queryFn: async () => {
      const response = await axios.get(`/today/data?location=${location}&lang=en`);
      if (response.data?.cod === "404") {
        throw new Error("City not found");
      }
      return response.data;
    },
    enabled: !!location,
  });

  
    // Show alert if there is an error fetching data
    useEffect(() => {
        if (error) {
        showAlert("Error fetching data", "error");
        }
    }, [error, showAlert]);


  return { data };
}