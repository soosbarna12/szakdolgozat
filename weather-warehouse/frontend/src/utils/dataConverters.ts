import dayjs from "dayjs";
import { HistoricalDataTable, ServerHistoricalData } from "../types/historicalDataTable.type";
import { HistoricalLocationData } from "../contexts/HistoricalContext/HistoricalContext.type";
import { getDataOrNA } from "./getDataOrNA";

export function convertKelvinToCelsius(kelvin: number | null): number | string {
	if (kelvin === null) {
		return 'N/A';
	}
	return Math.round(kelvin - 273.15);
}

export function convertKelvinToFahrenheit(kelvin: number | null): number | string {
	if (kelvin === null) {
		return 'N/A';
	}
	return Math.round(((kelvin - 273.15) * 9) / 5 + 32);
}

export function convertCelsiusToFahrenheit(celsius: number | null): number | string {
	if (celsius === null) {
		return 'N/A';
	}
	return Math.round((celsius * 9) / 5 + 32);
}

export function convertFahrenheitToCelsius(fahrenheit: number | null): number | string {
	if (fahrenheit === null) {
			return 'N/A';
	}
	return Math.round((fahrenheit - 32) * 5 / 9);
}

export function convertTitleCase(str: string): string {
	return str.charAt(0).toUpperCase() + str.slice(1).toLowerCase();
}

export function convertServerHistoricalData(historicalData: ServerHistoricalData[], location: HistoricalLocationData): HistoricalDataTable[] {
	return historicalData.map((data) => ({
		lat: location.lat,
		lon: location.lon,
		locationKey: data.LocationKey,
		dateKey: data.DateKey,
		date: dayjs(data.FullDate).format("YYYY-MM-DD"),
		cityName: data.CityName,
		countryCode: data.CountryCode,
		temp: getDataOrNA(data.Temperature),
		maxTemp: getDataOrNA(data.MaxTemperature),
		minTemp: getDataOrNA(data.MinTemperature),
		humidity: getDataOrNA(data.Humidity),
		windSpeed: getDataOrNA(data.WindSpeed),
		precipitation: getDataOrNA(data.Precipitation),
		pressure: getDataOrNA(data.Pressure),
		cloudCover: getDataOrNA(data.CloudCover),
	}));
}
