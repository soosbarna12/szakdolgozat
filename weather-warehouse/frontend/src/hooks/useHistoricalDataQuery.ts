import { useQuery } from "@tanstack/react-query";
import { useEffect } from "react";
import { UseHistoricalDataQueryProps } from "../pages/HistoricalPage/HistoricalPage.type";
import { useAlert } from "../utils/AlertContext";
import axios from "../utils/axiosConfig";
import { convertServerHistoricalData } from "../utils/dataConverters";

export function useHistoricalDataQuery(props: Readonly<UseHistoricalDataQueryProps>) {
  const { showAlert } = useAlert();
  const { location, date } = props;
  
  const { data, error, isLoading } = useQuery({
    queryKey: ["historicalWeatherData", location, date],
    queryFn: async () => {
      const response = await axios.post(`/api/historical/historicalData`, {
        location,
        date
      });

      return convertServerHistoricalData(response.data, location)?.[0] ?? {};
    },
    enabled: !!(location && date),
  });

  useEffect(() => {
    if (error) {
      showAlert("Error fetching historical weather data", "error");
    }
  }, [error]);

  return { data, error, isLoading };
}
