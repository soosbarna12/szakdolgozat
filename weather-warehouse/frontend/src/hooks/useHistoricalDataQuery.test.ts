import '@testing-library/jest-dom';
import { waitFor } from '@testing-library/react';
import { useAlert } from '../utils/AlertContext';
import axios from '../utils/axiosConfig';
import { convertServerHistoricalData } from '../utils/dataConverters';
import { renderHookWithQueryClient } from '../utils/test/renderHookWithQueryClient';
import { useHistoricalDataQuery } from './useHistoricalDataQuery';

jest.mock('../utils/AlertContext', () => ({
  useAlert: jest.fn(),
}));

jest.mock('../utils/axiosConfig');

jest.mock('../utils/dataConverters', () => ({
  convertServerHistoricalData: jest.fn(),
}));

describe('useHistoricalDataQuery', () => {
  const mockShowAlert = jest.fn();
  const mockLocation = { name: 'New York', lat: 40.7128, lon: -74.006 };
  const mockDate = '2023-01-01';
  const mockResponseData = [{ temperature: 5, humidity: 80 }];
  const mockConvertedData = [{ temperature: 5, humidity: 80, location: mockLocation }];

  beforeEach(() => {
    jest.clearAllMocks();
    (useAlert as jest.Mock).mockImplementation(() => ({
      showAlert: mockShowAlert,
    }));
  });

  it('matches useHistoricalDataQuery hook snapshot', async () => {
    (axios.post as jest.Mock).mockReturnValue({ data: mockResponseData });
    (convertServerHistoricalData as jest.Mock).mockReturnValueOnce(mockConvertedData);

    const { result } = renderHookWithQueryClient(() =>
      useHistoricalDataQuery({ location: mockLocation, date: mockDate })
    );

    await waitFor(() => {
      expect(result.current.isLoading).toBe(false);
      expect(result.current.error).toBeNull();
    });

    expect(result.current).toMatchSnapshot();
  });

  it('fetches historical data successfully and converts it', async () => {
    (axios.post as jest.Mock).mockReturnValue({ data: mockResponseData });
    (convertServerHistoricalData as jest.Mock).mockReturnValueOnce(mockConvertedData);

    const { result } = renderHookWithQueryClient(() =>
      useHistoricalDataQuery({ location: mockLocation, date: mockDate })
    );

    await waitFor(() => {
      console.log('result.current', result.current);
      expect(result.current.data).toEqual(mockConvertedData[0]);
      expect(result.current.isLoading).toBe(false);
      expect(result.current.error).toBeNull();
    });
  });

  it('returns an empty object when no data is found', async () => {
    (axios.post as jest.Mock).mockResolvedValueOnce({ data: [] });
    (convertServerHistoricalData as jest.Mock).mockReturnValueOnce([]);

    const { result } = renderHookWithQueryClient(() =>
      useHistoricalDataQuery({ location: mockLocation, date: mockDate })
    );

    await waitFor(() => {
      expect(result.current.data).toEqual({});
      expect(result.current.isLoading).toBe(false);
      expect(result.current.error).toBeNull();
    });
  });

  it('does not fetch data when location or date is missing', async () => {
    const { result } = renderHookWithQueryClient(() =>
      useHistoricalDataQuery({ location: null, date: '' })
    );

    await waitFor(() => {
      expect(result.current.data).toBeUndefined();
      expect(result.current.isLoading).toBe(false);
      expect(result.current.error).toBeNull();
    });

    expect(axios.post).not.toHaveBeenCalled();
  });
});