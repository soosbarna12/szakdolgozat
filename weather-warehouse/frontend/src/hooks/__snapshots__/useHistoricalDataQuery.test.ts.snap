// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useHistoricalDataQuery matches useHistoricalDataQuery hook snapshot 1`] = `
{
  "data": {
    "humidity": 80,
    "location": {
      "lat": 40.7128,
      "lon": -74.006,
      "name": "New York",
    },
    "temperature": 5,
  },
  "error": null,
  "isLoading": false,
}
`;
